source/agora/node/Fee.d

```diff
@@ -0,0 +1,157 @@
+/*******************************************************************************
+
+    Contains the function to calculate the fees used to store the data
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.Fee;
+
+import agora.common.Amount;
+
+import std.math;
+
+/*******************************************************************************
+
+    Calculates the fee of transaction data to store
+
+    Params:
+        data_size = The size of the data
+        factor = The factor to calculate for the fee of transaction data
+
+    Returns:
+        A fee payable to data storage
+
+*******************************************************************************/
+
+public Amount calculateDataFee(ulong data_size, uint factor) pure nothrow @safe @nogc
+{
+    const ulong decimal = 100L;
+    ulong fee = cast(ulong)(
+        round(
+            (
+                exp((cast(double)data_size / cast(double)factor)) -
+                1.0
+            ) *
+            decimal
+        ) * 10_000_000L / decimal
+
+    );
```

Comment:
This is a landmine. We shouldn't use floating point for this. If we need a higher precision, use the same technique as in `Amount`, a fixed precision based on integer arithmetic.